Cython/Compiler/ExprNodes.py

```diff
@@ -3377,6 +3380,12 @@ def analyse_types(self, env):
                 if self.value.type is unicode_type and not self.value.may_be_none():
                     # value is definitely a unicode string and we don't format it any special
                     return self.value
+
+        if self.expr_text:
+            # strings are unicode in python, so we can safely do this
+            string_node = StringNode(self.pos, value=StringEncoding.BytesLiteral(self.expr_text.encode('utf-8')))
+            return AddNode(self.pos, operand1=string_node, operand2=self, operator='+', type=string_node.type, is_temp=True)
```

Comment:
Why is this not done in the parser? It builds a sequence of strings – I don't see a need for an additional `AddNode`.

Reply:
This was suggested by @da-woods [here](https://github.com/cython/cython/pull/3939#discussion_r539574402) I'm kind of new and still trying to understand how all this works, so I don't know the best way to do this :sweat_smile:

Reply:
I have been known to give bad advice.... Although I think this was still an improvement on the code that it replaced.

But yes, given that the `expr_text` is known as you parse it it's probably most efficient to handle it in the parser. Maybe if `p_f_string_expr` returned the index and a sequence instead (e.g. either `[expr_text, FormattedValueNode]` or `[FormattedValueNode]`)?

Reply:
> I have been known to give bad advice

New code means that we're solving a problem that we haven't solved before (at least not in this context). It's perfectly ok to try a couple of approaches until we understand the problem well enough to find a _good_ way to solve it. Every "bad" approach helps in making the "obvious" solution clearer, until we're able to see it (and then slap out foreheads, commit it, and start coding up our next "bad" approach for the next problem).

Reply:
> I have been known to give bad advice.

I didnt mean to blame it on you :sweat_smile:. I just wanted to point out where I got the code from so that scoder could have a look at the discussion. I should have worded that better, sorry!

> Maybe if p_f_string_expr returned the index and a sequence instead (e.g. either [expr_text, FormattedValueNode] or [FormattedValueNode])?

Will try that!

Reply:
Moved to the parser